hoomd/md/WallData.h

```diff
@@ -178,7 +176,7 @@ vecPtToWall(const PlaneWall& wall, const vec3<Scalar>& position, bool& inside)
     {
     vec3<Scalar> t = position;
     Scalar d = dot(wall.normal, t) - dot(wall.normal, wall.origin);
-    inside = (((d >= 0.0) && wall.inside) || ((d < 0.0) && !(wall.inside))) ? true : false;
```

Comment:
@joaander @b-butler 
I'm afraid I'm going to have to argue we actually don't do this. While the other day I was thinking of it as a matter of convenience, when actually seeing this bit of code getting removed it occurred to me that it introduces a real issue.

Removing the inside parameter will actually mean that there's no way to make our space open vs closed. If you simply flip the normal, you're going to have the 0 point be in both wall sections.
I keep using a wall with a Gaussian potential as my test case and for this instance you can imagine that while the forces would both evaluate to zero and not cause a problem, the energy would be twice what it should be for that particle.
I don't feel like the user should have to think about this edge case and try to figure out what will work out in the floating point math to move the origin just enough to not trigger this. If we keep this parameter internally but don't like the vagueness of "inside" for the user we could rename it at the python level to "open"/"closed" or something like that.

Reply:
What use-case would require that the user be able to configure open vs closed spaces? In either case, we should make the plane geometry consistent with the sphere and cylinder geometries. Either we include the surface of the shape on the inside or we don't. In practical simulations, it doesn't make any difference because 1e-16 is infinitesimally far away from 0.

With any potential that is finite at r=0 (and `r_extrap=0`), the particles can pass through the wall into the outside space where they will move freely. There will be a large jump in the force and energy as the particle crosses the wall why may cause the integration to explode. This is not a configuration users will likely wish to choose. They will either use a force that goes infinite at r=0 or set `r_extrap > 0` so that particles that move outside the wall will be pushed back in.

I still think that having two variables (the normal direction and the `inside` flag) to determine which side of the plane the particles interact with is confusing to users. There should be one way to set this.

In your example are you imagining two walls place with opposing normals that the particle is supposed to pass through without a jump in energy and force? What research use case would that be used for? Wouldn't it be cleaner to have a `two_sided` flag that evaluates one wall potential on the absolute value of the signed distance?

Reply:
> 
> 
> What use-case would require that the user be able to configure open vs closed spaces? In either case, we should make the plane geometry consistent with the sphere and cylinder geometries. Either we include the surface of the shape on the inside or we don't. In practical simulations, it doesn't make any difference because 1e-16 is infinitesimally far away from 0.
> 
There's a few use cases I suppose we were trying to cover by making these wall geometries very general and flexible. Perhaps the name wall is what is most confusing about this however. I think while these options could be confusing to some users on a first read-through the documentation, those who are looking for more complicated uses would find a complete implementation that allows them to construct a consistent field geometry. I'll illustrate this more through the rest of my response.

Wall as a name obviously makes the most sense when you're talking about potentials which go infinite at r=0. For these use case whether the space is open or closed makes *no difference* as you pointed out. We made a choice that perhaps feels arbitrary now that inside would be the closed space, although I think this is still the choice I'd make for two reasons:
1) It can have an easy geometric interpretation for 2/3 geometries
2) Without introducing a new keyword we also can make it clear that the space is open or closed (i.e. inside is always closed/includes the bounds of the geometry)
If for some reason you think that it makes more sense for inside to always be open, I don't have a problem with it, but it might perhaps make sense to take a quick poll since I would then say the intuitive answer is subjective and we should default to a majority opinion and obviously still document it.

> With any potential that is finite at r=0 (and `r_extrap=0`), the particles can pass through the wall into the outside space where they will move freely. There will be a large jump in the force and energy as the particle crosses the wall why may cause the integration to explode. This is not a configuration users will likely wish to choose. They will either use a force that goes infinite at r=0 or set `r_extrap > 0` so that particles that move outside the wall will be pushed back in.
> 
I perhaps need to go back and double check that the implementation hasn't changed. But my original intent of the `r_extrap` functionality was explicitly to extend a potential interaction through the wall smoothly. So in this case, the particle should pass through the wall and still be assigned the force evaluated at `r_extrap`. So it should not move freely unless V(`r_extrap`)/dr = 0. And I believe the potential energy of the particle was also assigned it's value in this extended picture. This is more a case that covers one wall potential being "two-sided". The whole point was essentially to avoid cases where you could end up with a large jump and to do so in as consistent of a way as possible. This certainly breaks down a bit as the particle crosses the box boundary, but that is unavoidable and hopefully possible to anticipate and either avoid or track via the ParticleData image.

I wrote the following and then realized that I think I implemented this wrong for this functionality! It was documented how it was coded, but to cover this use case, we would need to make a minor change covering this exact scenario. So glad we had this discussion and glad things are getting better testing. It's clearly my own logic that had a mismatch when coding one part and another.

>> This does bring up one case that I think I should perhaps code a test case for to be sure it works (assuming we agree on how it should work) regarding the combination of a potential that is finite at r=0 (and `r_extrap=0`), I'll list the components:
>> - A Gauss wall potential applied to *any* geometry
>> - `r_extrap=0`
>> - `inside=True`
>> - At any point on the wall geometry we should then have the exact value `epsilon` for the potential energy and 0 force.
>> - At any point "outside" the wall geometry ^^
>> However for `inside=False` I'm honestly not positive if there would be an issue. I don't think so because...
>> 
This is where I double checked myself before saying,  "I believe `r_extrap=0` is meant to still allow the whole space being open or closed to be overridden." What I coded is `r_extrap<=0` but I'm now making the argument for `r_extrap<0`. I think this is another case of it not making a difference for the vast majority of wall potentials, but I've clearly constructed an edge case that I'd argue could be used so I think it should be coded differently. This will of course need to cover the python side handling converting False to -1. rather than 0 which is probably why I didn't do this correctly the first time... a bug masquerading as simplicity.

> I still think that having two variables (the normal direction and the `inside` flag) to determine which side of the plane the particles interact with is confusing to users. There should be one way to set this.
> 
I'm definitely not going to say you're wrong on this. But I think if we set a default of `inside=True` (or False if you think it make us more compatible with computer graphics conventions) for planes users that don't need this functionality would ignore it and users that do could still easily find, understand, and use it.

> In your example are you imagining two walls place with opposing normals that the particle is supposed to pass through without a jump in energy and force? What research use case would that be used for? Wouldn't it be cleaner to have a `two_sided` flag that evaluates one wall potential on the absolute value of the signed distance?
>
Revisiting my above example of test case. If we did want this Gauss wall potential to be two sided we could only correctly construct it by having access to control over the spaces being open and closed. And I think this is the clearest case for why these functionalities should be combined. The two_sided potential is the only time when you'd want to have an open and a closed potential, but you'd also be combining the same geometry in a way that that needs to be named and I believe it's intuitively to refer to it as "inside and outside" which combined  makes the whole space defined.